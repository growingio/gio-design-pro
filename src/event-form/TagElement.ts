// import { Element } from '../types';

// export declare type TagElement = Element;
export interface TagElement {
  id: string;
  name: string;
  actions: string[];
  attrs: DocProps;
  definition: DocProps;
  screenshot: Screenshot;
  platforms: string[];
  docType: string;
  description: string;
  businessType: string;
  isSystem: Boolean;
  createdAt: string;
  updatedAt: string;
  creatorId: string;
  updaterId: string;
  creator: string;
  updater: string;
  patternMatched: Boolean;
  appVersion: string;
  sdkVersion: string;
}

export interface DocProps {
  urlScheme?: string;
  domain?: string;
  path?: string;
  query?: string;
  xpath?: string;
  index?: string;
  href?: string;
  content?: string;
  pg?: string;
  contentType?: string; // 'match_phrase' | '=';
}
export interface Screenshot {
  target?: string;
  viewport: string;
}
