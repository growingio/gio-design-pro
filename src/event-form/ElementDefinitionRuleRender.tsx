/* eslint-disable react/no-danger */
import React from 'react';
import { DocProps, TagElement } from './TagElement';
// import { LimitCondition } from './types';

const renderDefinitionRuleText = (props: {
  attrs: DocProps;
  repeatTag?: TagElement;
  // appType: AppType;
  limitCondition: DocProps;
}) => {
  const { attrs: originAttrs, repeatTag, limitCondition } = props;
  // console.warn('props', { originAttrs, limitCondition });

  // const isNative = appType === AppType.NATIVE;
  let attrs = originAttrs;
  let limit = limitCondition as DocProps;
  if (repeatTag) {
    attrs = repeatTag.attrs;
    limit = repeatTag.definition;
  }

  // let { content } = limit;
  // if (content && content.length > 30) {
  //   content = `${content.substring(0, 28)}...`;
  // }
  // const index = +limit.index;
  // App 的index 从0开始；展示时 统一从1开始

  const Cel: React.FC<{}> = ({ children }) => (
    <span style={{ color: '#3867f4' }} className="link">
      {children}
    </span>
  );
  const renderAttrsDesc = () => {
    // let hasLimiter = !!get(data, 'definition.content') || !!get(data, 'definition.href');
    const contentDefined = !!limit.content;
    const indexDefined = !!limit.index;
    const hrefDefined = !!limit.href;
    const contentAttr = !!attrs?.content;
    const hrefAttr = !!attrs?.href;
    const indexAttr = !!attrs?.index;
    // const attrsDescs=['即使元素的跳转']
    // 现根据有无跳转链接判断同类和非同类的描述
    let str1: string = '';
    let str2: string = '';
    if (indexAttr) {
      if (contentAttr && hrefAttr && contentDefined && hrefDefined && indexDefined) {
        str1 = '内容、位置或跳转链接';
        str2 = '';
      }
      if (contentAttr && hrefAttr && !contentDefined && hrefDefined && indexDefined) {
        str1 = '位置或跳转链接';
        str2 = '内容';
      }
      if (contentAttr && hrefAttr && contentDefined && !hrefDefined && indexDefined) {
        str1 = '跳转链接';
        str2 = '内容和位置';
      }
      if (contentAttr && hrefAttr && !contentDefined && !hrefDefined && indexDefined) {
        str1 = '位置';
        str2 = '内容和跳转链接';
      }

      if (contentAttr && hrefAttr && !contentDefined && hrefDefined && !indexDefined) {
        str1 = '跳转链接';
        str2 = '内容和位置';
      }
      if (contentAttr && hrefAttr && contentDefined && !hrefDefined && !indexDefined) {
        str1 = '内容';
        str2 = '位置和跳转链接';
      }

      if (contentAttr && hrefAttr && contentDefined && hrefDefined && !indexDefined) {
        str1 = '内容或跳转链接';
        str2 = '位置';
      }
      if (contentAttr && hrefAttr && !contentDefined && !hrefDefined && !indexDefined) {
        str1 = '';
        str2 = '';
      }
      // 元素attrs中没有的属性 definition中肯定没有
      // 只有href
      if (!contentAttr && hrefAttr && hrefDefined && indexDefined) {
        str1 = '位置或跳转链接';
        str2 = '';
      }
      if (!contentAttr && hrefAttr && !hrefDefined && indexDefined) {
        str1 = '位置';
        str2 = '跳转链接';
      }
      if (!contentAttr && hrefAttr && !hrefDefined && !indexDefined) {
        str1 = '';
        str2 = '';
      }
      if (!contentAttr && hrefAttr && hrefDefined && !indexDefined) {
        str1 = '跳转链接';
        str2 = '位置';
      }
      // 只有content时
      if (contentAttr && !hrefAttr && !contentDefined && !indexDefined) {
        str1 = '';
        str2 = '';
      }
      if (contentAttr && !hrefAttr && contentDefined && !indexDefined) {
        str1 = '内容';
        str2 = '位置';
      }
      if (contentAttr && !hrefAttr && !contentDefined && indexDefined) {
        str1 = '位置';
        str2 = '内容';
      }
      if (contentAttr && !hrefAttr && contentDefined && indexDefined) {
        str1 = '内容或位置';
        str2 = '';
      }
      if (!contentAttr && !hrefAttr) {
        str1 = '';
        str2 = '';
      }
    } else {
      if (contentAttr && hrefAttr && contentDefined && hrefDefined) {
        str1 = '内容或跳转链接';
        str2 = '';
      }
      if (contentAttr && hrefAttr && !contentDefined && hrefDefined) {
        str1 = '跳转链接';
        str2 = '内容';
      }
      if (contentAttr && hrefAttr && contentDefined && !hrefDefined) {
        str1 = '内容';
        str2 = '跳转链接';
      }
      if (contentAttr && hrefAttr && !contentDefined && !hrefDefined) {
        str1 = '';
        str2 = '内容和跳转链接';
      }
      // --只有href
      if (!contentAttr && hrefAttr && !hrefDefined) {
        str1 = '';
        str2 = '跳转链接';
      }
      if (!contentAttr && hrefAttr && hrefDefined) {
        str1 = '跳转链接';
        str2 = '';
      }
      // --只有content
      if (contentAttr && !hrefAttr && !contentDefined) {
        str1 = '';
        str2 = '内容';
      }
      if (contentAttr && !hrefAttr && contentDefined) {
        str1 = '内容';
        str2 = '';
      }
      // 都没有
      if (!contentAttr && !hrefAttr) {
        str1 = '';
        str2 = '';
      }
    }
    const tmp = () => {
      if (!str1 && !str2 && indexAttr) {
        return <span>所有同类元素的数据之和。我们已经将这组同类元素用虚线框标记出来。</span>;
      }
      return (
        <>
          <span>{str1 ? `若元素的${str1}改变，就不再继续统计了。` : ''}</span>
          <span>{str2 ? `即使元素${str2}改变，也会继续统计。` : ''}</span>
        </>
      );
    };
    return tmp();
  };
  function ellipsisText(content?: string) {
    if (content && content.length > 30) {
      return `${content.substring(0, 28)}...`;
    }
    return content;
  }
  const renderSimilar = () => {
    const hasLimiter = !!limit.content || !!limit.href || limit.index;
    const hasContentLimiter = !!limit.content;
    const hasIndexLimiter = !!limit.index;
    const hasHrefLimiter = !!limit.href;
    return (
      <>
        {hasLimiter && (
          <>
            <Cel>同类元素内</Cel>
            {hasContentLimiter && hasIndexLimiter && (
              <>
                ，限定
                <Cel>
                  元素内容{limit.contentType === 'match_phrase' ? '包含' : '为'}「{ellipsisText(limit.content)}」
                </Cel>
                ，<Cel>{`元素位置为「第${limit.index}位」`}</Cel>
              </>
            )}
            {!hasContentLimiter && hasIndexLimiter && (
              <>
                ，限定<Cel>{`元素位置为「第${limit.index}位」`}</Cel>
              </>
            )}
            {hasContentLimiter && !hasIndexLimiter && (
              <>
                ，限定
                <Cel>
                  元素内容{limit.contentType === 'match_phrase' ? '包含' : '为'}「{ellipsisText(limit.content)}」
                </Cel>
              </>
            )}
            {hasHrefLimiter && (
              <>
                ，限定
                <Cel>跳转链接</Cel>
              </>
            )}
            的元素。
          </>
        )}
        {renderAttrsDesc()}
      </>
    );
  };

  const renderNoSimilar = () => {
    // let hasLimiter = !!get(data, 'definition.content') || !!get(data, 'definition.href');

    const hasContentLimiter = !!limit.content;
    // let hasIndexLimiter = get(data, 'definition.index') === null;
    const hasHrefLimiter = !!limit.href;
    return (
      <>
        <>
          <Cel>当前位置</Cel>
          {hasContentLimiter && (
            <>
              ，限定
              <Cel>
                元素内容{limit.contentType === 'match_phrase' ? '包含' : '为'}「{ellipsisText(limit.content)}」
              </Cel>
            </>
          )}
          {hasHrefLimiter && (
            <>
              ，限定
              <Cel>跳转链接</Cel>
            </>
          )}
          的元素。
        </>
        {renderAttrsDesc()}
      </>
    );
  };
  const hasSimilar = () => !!attrs?.index;

  return (
    <span>
      {repeatTag && (
        <span>
          该规则已被 <b>{repeatTag.creator}</b> 定义为
          <span>【{repeatTag.name}】。</span>
        </span>
      )}
      现在定义的是 <Cel>所属页面</Cel> 中，
      {hasSimilar() ? renderSimilar() : renderNoSimilar()}
    </span>
  );
};

export default renderDefinitionRuleText;
