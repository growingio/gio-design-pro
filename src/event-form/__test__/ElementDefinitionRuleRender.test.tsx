/* eslint-disable no-console */
import React from 'react';
// import { fireEvent, render, screen } from '@testing-library/react';
import { render, screen } from '@testing-library/react';
import ElementDefinitionRuleRender from '../ElementDefinitionRuleRender';
import { TagElement } from '../TagElement';
// import { mount, shallow, render } from 'enzyme';

jest.useFakeTimers();

describe('ElementDefinitionRuleRender', () => {
  it('can render with content ,index,href,', async () => {
    const attrs = {
      domain: 'release-messages.growingio.cn',
      path: 'pages/index/index',
      // query: undefined,
      xpath: '/div.title/form.basic-grey/h1/span',
      index: '1',
      href: '/link',
      content: '元素内容',
    };
    const limitCondition = {
      domain: 'release-messages.growingio.cn',
      path: 'pages/index/index',
      // query: undefined,
      // xpath: '/div.title/form.basic-grey/h1/span',
      index: '1',
      href: '/link',
      content: '元素内容',
      contentType: '=',
    };

    const wrapper = render(<ElementDefinitionRuleRender attrs={attrs} limitCondition={limitCondition} />);

    expect(wrapper).toMatchSnapshot();
    expect(screen.queryByText('若元素的内容、位置或跳转链接改变，就不再继续统计了。')).toBeTruthy();
  });
  it('render without content', async () => {
    const attrs = {
      domain: 'release-messages.growingio.cn',
      path: 'pages/index/index',
      // query: undefined,
      xpath: '/div.title/form.basic-grey/h1/span',
      index: '1',
      href: '/link',
      content: '元素内容',
    };
    const limitCondition = {
      domain: 'release-messages.growingio.cn',
      path: 'pages/index/index',
      // query: undefined,
      // xpath: '/div.title/form.basic-grey/h1/span',
      index: '1',
      href: '/link',
      // content: '元素内容',
      contentType: '=',
    };
    const wrapper = render(<ElementDefinitionRuleRender attrs={attrs} limitCondition={limitCondition} />);
    expect(wrapper).toMatchSnapshot();
    expect(screen.queryByText('若元素的位置或跳转链接改变，就不再继续统计了。')).toBeTruthy();
  });
  it('render without content,href,index', async () => {
    const attrs = {
      domain: 'release-messages.growingio.cn',
      path: 'pages/index/index',
      // query: undefined,
      xpath: '/div.title/form.basic-grey/h1/span',
      index: '1',
      href: '/link',
      content: '元素内容',
    };
    const limitCondition = {
      domain: 'release-messages.growingio.cn',
      path: 'pages/index/index',
      // query: undefined,
      // xpath: '/div.title/form.basic-grey/h1/span',
      // index: '1',
      // href: '/link',
      // content: '元素内容',
      contentType: '=',
    };
    const wrapper = render(<ElementDefinitionRuleRender attrs={attrs} limitCondition={limitCondition} />);
    expect(wrapper).toMatchSnapshot();
    expect(screen.queryByText('所有同类元素的数据之和。我们已经将这组同类元素用虚线框标记出来。')).toBeTruthy();
  });
  it('render with repeatTag', async () => {
    const attrs = {
      domain: 'release-messages.growingio.cn',
      path: 'pages/index/index',
      // query: undefined,
      // xpath: '/div.title/form.basic-grey/h1/span',
      index: '',
      href: '/link',
      content: '元素内容',
    };
    const limitCondition = {
      domain: 'release-messages.growingio.cn',
      path: 'pages/index/index',
      // query: undefined,
      // xpath: '/div.title/form.basic-grey/h1/span',
      index: '',
      href: '/link',
      content: '元素内容',
      contentType: '=',
    };
    const repeatTag = {
      id: 'kqQe1qQr',
      name: '定义一个新的事件2',
      creator: 'zhangxueqi',
      creatorId: 'oNGzxQgB',
      createdAt: '2021-01-08T06:21:14Z',
      updater: 'zhangxueqi',
      updaterId: 'oNGzxQgB',
      updatedAt: '2021-01-08T06:21:14Z',
      description: '',
      platforms: ['minp'],
      docType: 'elem',
      actions: ['clck'],
      patternMatched: false,
      attrs: {
        content: '主页',
        index: '0',
        href: 'pages/index/index',
        path: 'pages/index/index',
        // query: null,
        xpath: '#onTabItemTap',
        // contentType: null,
        domain: 'wx265d0fa6fa70fae9',
        // urlScheme: null,
        __typename: 'DocProps',
      },
      definition: {
        domain: 'wx265d0fa6fa70fae9',
        path: 'pages/index/index',
        query: '12=12',
        xpath: '#onTabItemTap',
        index: '0',
        href: 'pages/index/index',
        content: '主页',
        // pg: null,
        contentType: '=',
        // urlScheme: null,
        __typename: 'DocProps',
      },
      screenshot: { target: '', viewport: '', __typename: 'Screenshot' },
    };
    const wrapper = render(
      <ElementDefinitionRuleRender
        repeatTag={(repeatTag as unknown) as TagElement}
        attrs={attrs}
        limitCondition={limitCondition}
      />
    );
    expect(wrapper).toMatchSnapshot();
    // const exp = (
    //   <span>
    //     该规则已被
    //     <b>zhangxueqi</b>
    //     定义为
    //     <span>【 定义一个新的事件2 】。</span>
    //   </span>
    // );
    expect(screen.findByText(/^该规则已被/)).toBeTruthy();
  });
});
