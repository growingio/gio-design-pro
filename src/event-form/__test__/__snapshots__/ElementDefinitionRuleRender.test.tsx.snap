// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElementDefinitionRuleRender can render with content ,index,href, 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span>
        现在定义的是 
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          所属页面
        </span>
         中，
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          同类元素内
        </span>
        ，限定
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          元素内容
          为
          「
          元素内容
          」
        </span>
        ，
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          元素位置为「第1位」
        </span>
        ，限定
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          跳转链接
        </span>
        的元素。
        <span>
          若元素的内容、位置或跳转链接改变，就不再继续统计了。
        </span>
        <span />
      </span>
    </div>
  </body>,
  "container": <div>
    <span>
      现在定义的是 
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        所属页面
      </span>
       中，
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        同类元素内
      </span>
      ，限定
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        元素内容
        为
        「
        元素内容
        」
      </span>
      ，
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        元素位置为「第1位」
      </span>
      ，限定
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        跳转链接
      </span>
      的元素。
      <span>
        若元素的内容、位置或跳转链接改变，就不再继续统计了。
      </span>
      <span />
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ElementDefinitionRuleRender render with repeatTag 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span>
        <span>
          该规则已被 
          <b>
            zhangxueqi
          </b>
           定义为
          <span>
            【
            定义一个新的事件2
            】。
          </span>
        </span>
        现在定义的是 
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          所属页面
        </span>
         中，
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          同类元素内
        </span>
        ，限定
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          元素内容
          为
          「
          主页
          」
        </span>
        ，
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          元素位置为「第0位」
        </span>
        ，限定
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          跳转链接
        </span>
        的元素。
        <span>
          若元素的内容、位置或跳转链接改变，就不再继续统计了。
        </span>
        <span />
      </span>
    </div>
  </body>,
  "container": <div>
    <span>
      <span>
        该规则已被 
        <b>
          zhangxueqi
        </b>
         定义为
        <span>
          【
          定义一个新的事件2
          】。
        </span>
      </span>
      现在定义的是 
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        所属页面
      </span>
       中，
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        同类元素内
      </span>
      ，限定
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        元素内容
        为
        「
        主页
        」
      </span>
      ，
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        元素位置为「第0位」
      </span>
      ，限定
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        跳转链接
      </span>
      的元素。
      <span>
        若元素的内容、位置或跳转链接改变，就不再继续统计了。
      </span>
      <span />
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ElementDefinitionRuleRender render without content 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span>
        现在定义的是 
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          所属页面
        </span>
         中，
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          同类元素内
        </span>
        ，限定
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          元素位置为「第1位」
        </span>
        ，限定
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          跳转链接
        </span>
        的元素。
        <span>
          若元素的位置或跳转链接改变，就不再继续统计了。
        </span>
        <span>
          即使元素内容改变，也会继续统计。
        </span>
      </span>
    </div>
  </body>,
  "container": <div>
    <span>
      现在定义的是 
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        所属页面
      </span>
       中，
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        同类元素内
      </span>
      ，限定
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        元素位置为「第1位」
      </span>
      ，限定
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        跳转链接
      </span>
      的元素。
      <span>
        若元素的位置或跳转链接改变，就不再继续统计了。
      </span>
      <span>
        即使元素内容改变，也会继续统计。
      </span>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ElementDefinitionRuleRender render without content,href,index 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span>
        现在定义的是 
        <span
          class="link"
          style="color: rgb(56, 103, 244);"
        >
          所属页面
        </span>
         中，
        <span>
          所有同类元素的数据之和。我们已经将这组同类元素用虚线框标记出来。
        </span>
      </span>
    </div>
  </body>,
  "container": <div>
    <span>
      现在定义的是 
      <span
        class="link"
        style="color: rgb(56, 103, 244);"
      >
        所属页面
      </span>
       中，
      <span>
        所有同类元素的数据之和。我们已经将这组同类元素用虚线框标记出来。
      </span>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
